### PR TITLE
implement lookup arguments from logarithmic derivatives in pil

### PR DESCRIPTION
**Pull request by onurinanc**
This PR solves #1374

There are 3 examples included in the PR:

1. `lookup_via_challenges_ext_simple.asm` is the most basic example which implements {a} in {b} using extension field without using selectors
2. `lookup_via_challenges_ext_no_selector.asm` implements {a1, a2} in {b1, b2} using extension field without using selectors
3. `lookup_via_challenges_ext.asm` is a more complex example than others which implements {a1, a2, a3} in {b1, b2, b3} using extension field (which also handles tuples using Reed-Solomon fingerprinting). It also use different lhs and rhs selectors.

### CODE REVIEW

**Review by georgwiese — COMMENTED** [review]
Cool! 🥳 

I think we could look into how we can share code with the permutation argument, but I also think this can be a follow-up PR.

Few notes on tests:
- They are not running in the CI yet. You can add them [here](https://github.com/powdr-labs/powdr/blob/main/pipeline/tests/powdr_std.rs#L90).
- It seems like `lookup_via_challenges_ext_no_selector.asm` isn't testing anything the others aren't, right?
- I think in all your tests, all elements of your tuples are the same. So, for example, if the fingerprinting wouldn't respect the order, we wouldn't notice!
- There is no test for the non-extension case. I think it would be good to add at least one. I changed the tests locally actually to look at the compiled PIL, which is much easier to read in the non-extension case!

File: std/protocols/lookup.asm [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    let (update_expr_1, update_expr_2) = unpack_ext(
+        sub_ext(
+            add_ext(
+                mul_ext(
+                    mul_ext(lhs_folded, rhs_folded),
+                    sub_ext(next_acc, acc_ext)
+                ),
+                mul_ext(
+                    mul_ext(m_ext, from_base(rhs_selector)),
```
[the remark this answers is not in the document]
Comment (onurinanc) [reply std/protocols/lookup.asm:148]: So, for now, I'm leaving it this way

File: pipeline/tests/powdr_std.rs [hunk 2]
```diff
@@ -115,6 +115,38 @@ fn permutation_via_challenges_ext() {
...
+    test_halo2(f, Default::default());
+}
+
+#[test]
+fn lookup_via_challenges_ext() {
+    let f = "std/lookup_via_challenges_ext.asm";
+    test_halo2(f, Default::default());
+    // Note that this does not actually run the second-phase witness generation, because no
+    // Goldilocks backend support challenges yet. But at least it tests that the panic from
+    // the previous test is not happening.
```
Comment (georgwiese) [inline pipeline/tests/powdr_std.rs:130]: I think the "previous test" reference doesn't apply here, also in the next test.

File: std/protocols/lookup.asm [hunk 3]
```diff
@@ -52,49 +53,45 @@ let<T: Add + Mul + FromLiteral> compress_expression_array: T[], Fp2<T> -> Fp2<T>
...
+        Fp2::Fp2(a0_fe, a1_fe) => [a0_fe, a1_fe]
+    }
+};
     
 // Adds constraints that enforce that rhs is the lookup for lhs
 // Arguments:
 // - acc: A phase-2 witness column to be used as the accumulator. If 2 are provided, computations
 //        are done on the F_{p^2} extension field.
 // - lookup_constraint: The lookup constraint
+// - multiplicities: The multiplicities which shows how much time looked-up values (LHS) exists in the lookup table (RHS)
```
Comment (georgwiese) [inline std/protocols/lookup.asm:94]:
```suggestion
// - multiplicities: The multiplicities which shows how many times each RHS value appears in the LHS
```

**Review by georgwiese — COMMENTED** [review]
Cool! Just two small comments.

File: std/protocols/lookup.asm (hunk 3 above)
Comment (onurinanc) [reply std/protocols/lookup.asm:94]: Yes, that's better

**Review by georgwiese — APPROVED** [review]
🚀

**Pull request merged**